cassiopeia/simulator/Cas9LineageTracingDataSimulator.py

```diff
@@ -0,0 +1,367 @@
+    function to simulate Cas9-based mutations onto a defined set of "cassette".
+    These cassettes emulate the TargetSites described in Chan et al, Nature 2019
+    and the GESTALT arrays described in McKenna et al, Science 2020 in which
+    several Cas9 cut-sites are arrayed together. In the Chan et al technology,
+    these cassettes are of length 3; in McKenna et al, these cassettes are of
+    length 10.
+
+    The class accepts several parameters that govern the Cas9-based tracer.
+    First and foremost is the cutting rate, describing how fast Cas9 is able to
+    cut. We model Cas9 cutting as an exponential process, parameterized by the
+    specified mutation rate - specifically, for a lifetime `t` and parameter
+    `lambda`, the expected probability of Cas9 mutation, per site, is
+    `exp(-lambda * t)`.
+    
+    Second, the class accepts the architecture of the recorder - described by
+    the size of the cassette (by default 3) and the number of cassettes. The
+    resulting lineage will have (size_of_cassette * number_of_cassettes)
+    characters. The architecture is important in the sense that it encodes
+    the correlation between characters. This is critical in two ways: the
+    first is with respect to silencing -- entire cassettes are lost from
+    silencing, either transcriptional or stochastic events. The second is with
+    respect to Cas9 resections, in which more than one Cas9 molecule introduces
+    a cut to the same cassette. In this event, all cut sites intermediate will
+    be observed as missing. In this simulation, we allow this event to occur
+    if Cas9 molecules perform cuts on the same cassette at any point in a cell's
+    lifetime. Importantly, setting the cassette length to 1 will remove any
+    resection events due to Cas9 cutting, and will reduce the amount of
+    transcriptionally silencing observed.
+
+    Third, the class accepts a state distribution describing the relative
+    likelihoods of various indels. This is very useful, as it is typical that a
+    handful of mutations are far likelier than the bulk of the possible
+    mutations.
+
+    Finally, the class accepts two types of silencing rates. The first is the
+    heritable silencing rate which is a rare event in which an entire cassette
+    is transcriptionally silenced and therefore not observed. The second type
+    of silencing is a stochastic dropout rate which simulates the loss of
+    cassettes due to the low sensitivity of the RNA-sequencing assay. 
+
+    The function `overlay_data` will operate on the tree in place and will
+    specifically modify the data stored in the character attributes.
+
+    Args:
+        number_of_cassettes: Number of cassettes (i.e., arrays of target sites)
+        size_of_cassette: Number of editable target sites per cassette
+        mutation_rate: Exponential parameter for the Cas9 cutting rate. The user
+            can optionally pass in a float, in which every site will mutate at
```

Comment:
I don't think this is `optionally`, I would say `either` instead?

Reply:
This is what I meant in this context; happy to change it to "either" for clarity

Reply:
Oh, I see! I was thinking in the context of python types: `mutation_rate` is a `Union` rather than an `Optional`. I'm glad to leave as it is though, I'm just overloaded with mypy!

Reply:
Haha, no worries! I changed it to "either" for clarity.